### PR TITLE
Document pagination API

### PR DESCRIPTION
**Pull request by iilyak**
## Overview

Documentation for pagination API.

## Testing recommendations

`make check` should be sufficient

## Related Pull Requests

* [RFC-014 : Add pagination proposal](https://github.com/apache/couchdb-documentation/pull/530)
* [apache/couchdb#2870 : Implement paginated API](https://github.com/apache/couchdb/pull/2870)
* [apache/couchdb#2896 : Fix handling of limit query parameter](https://github.com/apache/couchdb/pull/2896)
* [apache/couchdb#2904 : Add support for previous bookmark](https://github.com/apache/couchdb/pull/2904)

## Checklist

- [ ] Update [rebar.config.script](https://github.com/apache/couchdb/blob/master/rebar.config.script) with the commit hash once this PR is rebased and merged

### CODE REVIEW

**Comment by wohali** [comment]
@iilyak can you freshen this PR please?

```
[2020-05-25T14:41:49.665Z] + make check

[2020-05-25T14:41:49.665Z] python3 ext/linter.py src/

[2020-05-25T14:41:49.665Z] /home/jenkins/workspace/cm1_CouchDB-Documentation_PR-550/src/ddocs/views/pagination.rst

[2020-05-25T14:41:49.665Z]   line 127 : too many empty lines

[2020-05-25T14:41:49.665Z]   line 178 : too many empty lines

[2020-05-25T14:41:49.665Z] /home/jenkins/workspace/cm1_CouchDB-Documentation_PR-550/src/api/ddoc/views.rst

[2020-05-25T14:41:49.665Z]   line 948 : too many empty lines

[2020-05-25T14:41:49.665Z] make: *** [Makefile:48: check] Error 1

script returned exit code 2
```

File: src/api/database/bulk-api.rst [hunk 1]
```diff
@@ -332,6 +332,9 @@ Sending multiple queries to a database
 
     :param db: Database name
 
+    :query number page_size: Specify the number of queries in the result.
```
Comment (flimzy) [inline src/api/database/bulk-api.rst:335]: For consistency with the other descriptions, I wouldn't make this a complete sentence.

```suggestion
    :query number page_size: The number of queries in the result.
```

File: src/api/ddoc/views.rst [hunk 2]
```diff
@@ -28,6 +28,7 @@
     :<header Accept: - :mimetype:`application/json`
                      - :mimetype:`text/plain`
 
+    :query string bookmark: Specify a bookmark to get sepcific page of the results.
```
Comment (flimzy) [inline src/api/ddoc/views.rst:31]:
```suggestion
    :query string bookmark: A bookmark to get sepcific page of the results.
```

File: src/api/ddoc/views.rst [hunk 3]
```diff
@@ -60,6 +61,8 @@
       the keys specified in the array.
     :query number limit: Limit the number of the returned documents to the
       specified number.
+    :query number page_size: Specify the number of rows in the result.
```
Comment (flimzy) [inline src/api/ddoc/views.rst:64]:
```suggestion
    :query number page_size: The number of rows in the result.
```

File: src/api/ddoc/views.rst [hunk 4]
```diff
@@ -914,3 +917,45 @@ Sending multiple queries to a view
...
     }
+
+.. _api/ddoc/view/pagination:
+
+Pagination
+==========
+
+.. versionadded:: 4.0
+
+There is a support for token based pagination. The token based pagination is an
```
Comment (flimzy) [inline src/api/ddoc/views.rst:928]: [Stylistic] Avoid passive voice, and combine sentences to reduce repetitiveness.

```suggestion
CouchDB supports token-based pagination, which is an
```

File: src/api/ddoc/views.rst [hunk 5]
```diff
@@ -914,3 +917,45 @@ Sending multiple queries to a view
...
+alternative to previously recomended :ref:`pagination recipe <views/pagination>`
```
Comment (flimzy) [inline src/api/ddoc/views.rst:929]: [Stylistic] Whether it was a previous recommendation probably isn't relevant here. Might be good detail in a changelog, though.

```suggestion
alternative to :ref:`pagination recipe <views/pagination>`
```

File: src/api/ddoc/views.rst [hunk 6]
```diff
@@ -914,3 +917,45 @@ Sending multiple queries to a view
...
+based on ``offset``, ``limit`` and ``skip``.
+
+The user can request paginated mode by setting ``page_size`` query parameter. When
+pagination is enabled the response would include ``next`` and ``previous`` tokens.
+Which can be used to retrieve next and previous page of the results.
```
Comment (flimzy) [inline src/api/ddoc/views.rst:934]:
```suggestion
The user can request paginated mode by setting the ``page_size`` query parameter. When
pagination is enabled, the response will include ``next`` and ``previous`` tokens,
which can be used to retrieve the next and previous pages of results, respectively.
```

File: src/api/ddoc/views.rst [hunk 7]
```diff
@@ -914,3 +917,45 @@ Sending multiple queries to a view
...
+The maximum possible page size is configured via ``request_limits``
```
Comment (flimzy) [inline src/api/ddoc/views.rst:935]: The above description indicates that "page_size" controls the maximum possible page size, so I gather this means something else. Is this referring to the maximum number of pages?  If so, I suggest rewording to make that more clear:

```suggestion
The maximum number of pages is configured via ``request_limits``
```

If it's something else, can you clarify?

File: src/api/ddoc/views.rst [hunk 8]
```diff
@@ -914,3 +917,45 @@ Sending multiple queries to a view
...
+    [request_limits]
+    _all_docs = 5000
+    _all_docs/queries = 5000
+    _all_dbs = 5000
+    _view = 2500
+    _view/queries = 2500
+
+
+Note that ``page_size`` for :ref:`Multiple queries <api/ddoc/view/multiple_queries>`
+endpoints limits number of queries the user can submit in the body of the request.
```
Comment (flimzy) [inline src/api/ddoc/views.rst:949]:
```suggestion
endpoints limits the number of queries the user can submit in the body of the request.
```

Comment (flimzy) [reply src/api/ddoc/views.rst:949]: How is the `page_size` limit related to the maximum queries a user can submit?  Is it a 1:1 relationship?  (i.e. page_size=2 means 2 queries?)

File: src/api/ddoc/views.rst [hunk 9]
```diff
@@ -914,3 +917,45 @@ Sending multiple queries to a view
...
+
+Compatibility notes
+-------------------
+
+- ``page_size`` is forbidden in the query object passed in ``queries`` array \
+  submitted via :post:`/{db}/_design/{ddoc}/_view/{view}/queries` request.
+
+- ``keys`` propery is incompatible with ``page_size``.
```
Comment (flimzy) [inline src/api/ddoc/views.rst:957]:
```suggestion
- The ``keys`` property is incompatible with ``page_size``.
```

File: src/api/ddoc/views.rst [hunk 10]
```diff
@@ -914,3 +917,45 @@ Sending multiple queries to a view
...
+
+- value for ``skip`` property has to be in range of ``[0..<request_limit>]``.
```
Comment (flimzy) [inline src/api/ddoc/views.rst:959]: [Stylistic]

```suggestion
- value for ``skip`` property must be in range of ``[0..<request_limit>]``.
```

File: src/ddocs/views/pagination.rst [hunk 11]
```diff
@@ -123,8 +123,60 @@ Or in a pseudo-JavaScript snippet:
         page.display_link('next');
     }
 
-Paging
-======
+
+Paging (New Way)
```
Comment (flimzy) [inline src/ddocs/views/pagination.rst:127]: Could we give these proper names, rather than simply "old" vs "new"?

Perhaps the old way could be called "View-based Paging" for example?

File: src/ddocs/views/pagination.rst [hunk 12]
```diff
@@ -123,8 +123,60 @@ Or in a pseudo-JavaScript snippet:
...
+        {"id":"a0746072bba60a62b01209f467ca4fe2","key":"Biffy Clyro","value":null},
+        {"id":"b47d82284969f10cd1b6ea460ad62d00","key":"Foo Fighters","value":null},
+        {"id":"45ccde324611f86ad4932555dea7fce0","key":"Tenacious D","value":null},
+        {"id":"d7ab24bb3489a9010c7d1a2087a4a9e4","key":"Future of the Left","value":null},
+        {"id":"ad2f85ef87f5a9a65db5b3a75a03cd82","key":"Helmet","value":null}
+    ],
+        "next": "an encoded string representing bookmark pointing to next page of results"
+    }
+
+By presence of ``next`` property we can determine if there  are more pages to display.
```
Comment (flimzy) [inline src/ddocs/views/pagination.rst:151]:
```suggestion
By presence of ``next`` property we can determine if there are more pages to display.
```

File: src/ddocs/views/pagination.rst [hunk 13]
```diff
@@ -123,8 +123,60 @@ Or in a pseudo-JavaScript snippet:
...
+
+To get next page from CouchDB we would use::
```
Comment (flimzy) [inline src/ddocs/views/pagination.rst:153]:
```suggestion
To get the next page from CouchDB we would use::
```

File: src/ddocs/views/pagination.rst [hunk 14]
```diff
@@ -123,8 +123,60 @@ Or in a pseudo-JavaScript snippet:
...
+
+    curl -X GET 'http://127.0.0.1:5984/artists/_design/artists/_view/by-name?bookmark=<the next token>'
+
+The result:
```
Comment (flimzy) [inline src/ddocs/views/pagination.rst:157]:
```suggestion
The result::
```

File: src/ddocs/views/pagination.rst [hunk 15]
```diff
@@ -123,8 +123,60 @@ Or in a pseudo-JavaScript snippet:
...
+        {"id":"3e1b84630c384f6aef1a5c50a81e4a34","key":"Perfect Circle","value":null},
+        {"id":"84a371a7b8414237fad1b6aaf68cd16a","key":"Queens of the Stone Age",
+        "value":null},
+        {"id":"dcdaf08242a4be7da1a36e25f4f0b022","key":"Silverchair","value":null}
+    ],
+        "previous": "an encoded string representing bookmark pointing to previous page of results",
+        "next": "an encoded string representing bookmark pointing to next page of results",
+    }
+
+The ``previous`` property can be used to get the previous page of the results if need to::
```
Comment (flimzy) [inline src/ddocs/views/pagination.rst:173]:
```suggestion
Likewise, the ``previous`` property can be used to get the previous page of the results::
```

File: src/api/ddoc/views.rst (hunk 8 above)
Comment (iilyak) [reply src/api/ddoc/views.rst:949]: >  Is it a 1:1 relationship? (i.e. page_size=2 means 2 queries?)

yes

File: src/api/ddoc/views.rst [hunk 16]
```diff
@@ -28,6 +28,7 @@
     :<header Accept: - :mimetype:`application/json`
                      - :mimetype:`text/plain`
 
+    :query string bookmark: A bookmark to get sepcific page of the results.
```
Comment (flimzy) [inline src/api/ddoc/views.rst:31]:
```suggestion
    :query string bookmark: A bookmark to get specific page of the results.
```

Comment (iilyak) [reply src/api/ddoc/views.rst]: fixed

**Comment by iilyak** [comment]
@flimzy I think I fixed all problems. Could you take a second look?

**Comment by wohali** [comment]
+1, let's merge it.

**Comment by wohali** [comment]
Ilya I am going to have to revert this because this is a 4.0 change, and we may want to get another 3.x release out the door before 4.0 hits.
